guest/RATIONALE.md

```diff
@@ -47,7 +47,38 @@ mean amplifying that overhead to a point where second or longer latency would
 be possible. For reasons like this, we dismissed Go 1.21 for TinyGo, and will
 revisit when Go 1.22 introduces [`//go:wasmexport`][3].
 
+## Why do we use nottinygc?
+
+Kubernetes Scheduler plugins need to access very large model data, specifically
+node and pod data. WebAssembly has a sandbox model, so the memory of a plugin
+is not the same as host scheduler process. This means the node and pod are not
+passed by reference. In fact they cannot be copied by value either. Instead,
+they must be marshaled to wire format and unmarshalled within the WebAssembly
+guest.
+
+Unmarshalling protobuf messages with default tooling creates a lot of garbage.
+As WebAssembly has only one thread, garbage collection is inlined. TinyGo's
+compiler optimized for code size not speed. In the default configuration, we
+found inlined GC overhead to be over half the latency of a plugin execution.
+
+[nottinygc][4] is an alternate garbage collection implementation for TinyGo.
+This optimizes for performance instead of bytecode size. Swapping this added
+110KB to the base size of the wasm guest, but resulted in 48pct drop in
+execution time of a simple plugin, in a scale of hundreds of microseconds.
+
+nottinygc has tradeoffs besides size. One is that it isn't built-in to TinyGo.
+To use nottinygc requires custom flags in the build process, in addition to the
+custom ones we already have. We expect end users to use our SDK scripts, to
+avoid the chance of missing these flags. We also added tests so that lack
+thereof will be easy to identify.
+
+Another tradeoff is that nottinygc is new and currently only one contributor.
```

Comment:
Think this is a fair section.

Reply:
also fair mostly to say same about WASI 😎